src/python_testing/TC_OCC_2_1.py

```diff
@@ -0,0 +1,225 @@
+                                "OccupancySensorType is not in valid range")
+        else:
+            logging.info("OccupancySensorType attribute is a mandatory attribute. Test step fails")
+
+        self.step(4)
+        if attributes.OccupancySensorTypeBitmap.attribute_id in attribute_list:
+            occupancy_sensor_type_bitmap_dut = await self.read_occ_attribute_expect_success(endpoint=endpoint, attribute=attributes.OccupancySensorTypeBitmap)
+            asserts.assert_less_equal(occupancy_sensor_type_bitmap_dut, 0b00000111, "OccupancySensorTypeBitmap attribute is not in valid range")
+        else:
+            logging.info("OccupancySensorTypeBitmap attribute is a mandatory attribute. Test step fails")
+
+        self.step(5)
+        if attributes.HoldTimeLimits.attribute_id in attribute_list:
+            if attributes.HoldTime.attribute_id in attribute_list: # check HoldTime conformance
+                hold_time_limits_dut = await self.read_occ_attribute_expect_success(endpoint=endpoint, attribute=attributes.HoldTimeLimits)
+
+                asserts.assert_less_equal(hold_time_limits_dut.HoldTimeMin, hold_time_limits_dut.HoldTimeMax, "HoldTimeMin is not in valid range")
+                asserts.assert_greater_equal(hold_time_limits_dut.HoldTimeMin, 0, "HoldTimeMin is not in valid range")
+                asserts.assert_less_equal(hold_time_limits_dut.HoldTimeMax, 0xFFFE, "HoldTimeMin is not in valid range")
+                asserts.assert_greater_equal(hold_time_limits_dut.HoldTimeMax, hold_time_limits_dut.HoldTimeMin, "HoldTimeMin is not in valid range")
+                asserts.assert_less_equal(hold_time_limits_dut.HoldTimeDefault, hold_time_limits_dut.HoldTimeMax, "HoldTimeMin is not in valid range")
+                asserts.assert_greater_equal(hold_time_limits_dut.HoldTimeDefault, hold_time_limits_dut.HoldTimeMin, "HoldTimeMin is not in valid range")
+            else:
+                logging.info("HoldTime conformance failed.  Test step skipped")
+
+        else:
+            logging.info("HoldTimeLimits not supported. Test step skipped")
+        
+        self.step(6)
+        if attributes.HoldTime.attribute_id in attribute_list:
+            hold_time_dut = await self.read_occ_attribute_expect_success(endpoint=endpoint, attribute=attributes.HoldTime)
+
+            asserts.assert_less_equal(hold_time_dut, 0xFFFE, "HoldTime attribute is out of range")
+            asserts.assert_greater_equal(hold_time_dut, 0, "HoldTime attribute is out of range")
+        else:
+            logging.info("HoldTime not supported. The rest of legacy attribute test can be skipped")
+
+        self.step(7)
+        if attributes.PIROccupiedToUnoccupiedDelay.attribute_id in attribute_list:
+            if ((occupancy_sensor_type_bitmap_dut == 0b00000001) | ((occupancy_sensor_type_bitmap_dut != 0b00000001)&(occupancy_sensor_type_bitmap_dut != 0b00000010)&(occupancy_sensor_type_bitmap_dut != 0b00000100))):

```

Comment:
Is this trying to assess if this is a PIR sensor? I think that's what this is doing per the test plan, but this will also catch 0b110 (ultrasonic and physical)

I think what you want is 
if (occupancy_sensor_type_bitmap_dut & Clusters.OccupancySensing.Enums.OccupancySensorTypeEnum.kPir) != 0:

Reply:
use "or" for boolean conditioning for clarity.
Also probably want constants for those bits. I cannot readily/easily read and review this long line as it stands.

Reply:
at least the `(occupancy_sensor_type_bitmap_dut != 0b00000001)` is redudant because if that is true, the first part matches and the first if is ok. Please simplify for readability.

Reply:
OccupancySensorType is an enum: 
![image](https://github.com/user-attachments/assets/188c41aa-fdff-405a-a389-01045ee9d9e9)


We should not do bitmap operations on enums. These are `==` or `in` tests.

Reply:
I.e. I am wondering why we use the bitmap here instead of the enum. Is this because things may be compound and an enum is insufficient?

Reply:
@andy31415 Followed Cecille's suggestion to make it maybe more readable as follow;  Still using bitmap for comparison though.

has_pir_bitmap = (occupancy_sensor_type_bitmap_dut & Clusters.OccupancySensing.Enums.OccupancySensorTypeEnum.kPir) == 1

has_ultrasonic_bitmap = (occupancy_sensor_type_bitmap_dut & Clusters.OccupancySensing.Enums.OccupancySensorTypeEnum.kUltrasonic) == 1

has_phy_bitmap = (occupancy_sensor_type_bitmap_dut & Clusters.OccupancySensing.Enums.OccupancySensorTypeEnum.kPhysicalContact) == 1

if (has_pir_bitmap == 1) or ((has_pir_bitmap == 0)&(has_ultrasonic_bitmap == 0)&(has_phy_bitmap == 0)):